experimental/ActionPolyRing/src/Content.jl

```diff
@@ -0,0 +1,1373 @@
+"""
+degrees(apre::ActionPolyRingElem) = map(i -> degree(apre, i), 1:ngens(parent(apre)))
+
+@doc raw"""
+    is_constant(p::ActionPolyRingElem)
+
+Return `true` if `p` is a degree zero polynomial or the zero polynomial, i.e. a constant polynomial. 
+"""
+is_constant(apre::ActionPolyRingElem) = is_constant(data(apre))
+
+@doc raw"""
+    vars(p::ActionPolyRingElem)
+
+Return the variables actually occuring in `p`. The variables are sorted with respect to the ranking of the action polynomial ring containing `p`, leading with the largest variable.
+"""
+vars(apre::ActionPolyRingElem) = sort(parent(apre).(vars(data(apre))); rev = true)
+
+is_gen(apre::ActionPolyRingElem) = is_gen(data(apre))
+
+@doc raw"""
+    gen(apr::ActionPolyRing, i::Int, midx::Vector{Int})
+
+Return the `i`-th elementary variable with multiindex `midx` in the action polynomial
+`apr`. If this jet variable was untracked, it is tracked afterwards. The index of the jet variable may also be passed as a tuple.
+
+# Examples
+
+```jldoctests
+julia> dpr = differential_polynomial_ring(ZZ, [:a, :b, :c], 4)[1]; gen(dpr, 1, [3,1,0,0])
+a[3,1,0,0]
+
+julia> gen(dpr, (1, [3,2,0,0]))
+a[3,2,0,0]
+
+julia> gens(dpr)
+5-element Vector{DifferentialPolyRingElem{ZZRingElem}}:
+ a[3,2,0,0]
+ a[3,1,0,0]
+ a[0,0,0,0]
+ b[0,0,0,0]
+ c[0,0,0,0]
+```
+"""
+function gen(apr::ActionPolyRing, i::Int, jet::Vector)
+  @req __is_valid_jet(apr, i, jet) "invalid jet variable"
+  jtv = __jtv(apr)
+  if haskey(jtv, (i, jet))
+    return deepcopy(jtv[(i, jet)])
+  end
+  __add_new_jetvar!(apr, i, jet)
+  return deepcopy(jtv[(i, jet)])
+end
+
+gen(apr::ActionPolyRing, jet_idx::Tuple{Int, Vector{Int}}) = gen(apr, jet_idx...)
+
+@doc raw"""
+    gen(apr::ActionPolyRing, i::Int)
+
+Return the, among the currently tracked variables of `apr`, 'i'-th largest one.
+
+# Examples
+
+```jldoctests
+julia> dpr = difference_polynomial_ring(ZZ, [:a, :b, :c], 4)[1]; gen(dpr, 2)
+b[0,0,0,0]
+
+julia> set_ranking!(dpr; partition = [[0,1,1],[1,0,0]]); gen(dpr, 2)
+c[0,0,0,0]
+```
+"""
+gen(apr::ActionPolyRing, i::Int) = gens(apr)[i]
+
+@doc raw"""
+    gens(apr::ActionPolyRing, jet_idxs::Vector{Tuple{Int, Vector{Int}}})
+
+Return the jet variables of the action polynomial ring `apr` specified by the entries of `jet_idxs` as
+a vector and track all new variables.
+
+# Examples
+
+```jldoctests
+julia> dpr = differential_polynomial_ring(ZZ, [:a, :b, :c], 4)[1]; gens(dpr, [(1, [3,1,0,0]), (1, [3,2,0,0])])
+2-element Vector{DifferentialPolyRingElem{ZZRingElem}}:
+ a[3,1,0,0]
+ a[3,2,0,0]
+
+julia> gens(dpr)
+5-element Vector{DifferentialPolyRingElem{ZZRingElem}}:
+ a[3,2,0,0]
+ a[3,1,0,0]
+ a[0,0,0,0]
+ b[0,0,0,0]
+ c[0,0,0,0]
+```
+"""
+function gens(apr::ActionPolyRing, jet_idxs::Vector{Tuple{Int, Vector{Int}}})
+  jtv = __jtv(apr)
+  new_jet_idxs = filter(jet_idx -> !haskey(jtv, jet_idx), jet_idxs)
+  if !is_empty(new_jet_idxs)
+    __add_new_jetvar!(apr, new_jet_idxs)
+  end
+  return map(jet_idx -> jtv[jet_idx], jet_idxs)
+end
```

Comment:
why not just
```suggestion
function gens(apr::ActionPolyRing, jet_idxs::Vector{Tuple{Int, Vector{Int}}})
  return [gen(apr, jet), for jet in jet_idxs]
end
```
? did you benchmark your implementation to be more efficient?

Reply:
The current implementations seems more efficient:

```julia
julia> dpr1, (a,b,c) = difference_polynomial_ring(ZZ, [:a, :b, :c], 4)
(Difference polynomial ring in 3 elementary symbols over ZZ, DifferencePolyRingElem{ZZRingElem}[a[0,0,0,0], b[0,0,0,0], c[0,0,0,0]])

julia> dpr, (a,b,c) = difference_polynomial_ring(ZZ, [:a, :b, :c], 4)
(Difference polynomial ring in 3 elementary symbols over ZZ, DifferencePolyRingElem{ZZRingElem}[a[0,0,0,0], b[0,0,0,0], c[0,0,0,0]])

julia> vec = [(rand(1:3), rand(0:10, 4)) for _ in 1:10000];

julia> @btime Oscar.gens1($dpr1, $vec);
  14.611 ms (90035 allocations: 46.74 MiB)

julia> @btime Oscar.gens($dpr, $vec);
  571.497 μs (9 allocations: 234.58 KiB)
 ```

Reply:
I see the difference now. You implementation is better if indices repeat in the list

Reply:
I thought about this again: you implementation is more efficient since you don't do deepcopies here. But you do them in `gens`, which is somehow inconsistent.
Of you want to protect your internal data structures from accidental modification from outside, you need to add deepcopy here as well. And then it should be basically the same as my initial suggestion

Reply:
```julia
julia> dpr, _ = difference_polynomial_ring(ZZ, 10, 5);

julia> dpr1, _ = difference_polynomial_ring(ZZ, 10, 5);

julia> vec = [(rand(1:10), rand(0:30, 5)) for _ in 1:20000];

julia> @btime Oscar.gens1($dpr1, $vec);
  38.446 ms (180024 allocations: 201.46 MiB)

julia> @btime Oscar.gens($dpr, $vec);
  62.232 ms (140009 allocations: 391.39 MiB)
```
Actually, yours is considerably faster now. I will implement it.